### PR TITLE
Improve Location Searching on Map

### PR DESCRIPTION
**Pull request by SofieDunt**
## Why

[Search goes to correct location](https://code4community-team.monday.com/boards/925293934/views/18904377/pulses/1398347345)
[Add pinpoint to searched location](https://code4community-team.monday.com/boards/925293934/views/18904377/pulses/1398349408)
[Predict search on enter](https://code4community-team.monday.com/boards/925293934/views/18904377/pulses/1114588342)

Improves the user's experience when searching locations on the map by making sure the map zooms to the location properly, adding a marker to clearly show the location requested, and predicting a location when a user hits enter before entering a complete address.

## This PR
Added logic to separate files where possible to keep the map view component readable. Tried to extract constants and logic from the useEffect where possible - this component will need more cleanup in the future.

## Screenshots
Marker

 

## Verification Steps
Tested searching locations on the map, autocomplete works when a prediction is available and does nothing when there are no predictions.

### CODE REVIEW

File: src/components/mapPageComponents/logic/view.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// Map logic for setting the view (zoom and position)
```
Comment (willmt80) [inline src/components/mapPageComponents/logic/view.ts:1]: 💯

File: src/components/mapPageComponents/logic/predict.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import { BOSTON_BOUNDS } from '../mapView';
+
+// Map logic for making predictions from user input
```
Comment (willmt80) [inline src/components/mapPageComponents/logic/predict.ts:3]: Is this logic all found within the Google API?

File: src/components/mapPageComponents/mapView/index.tsx [hunk 3]
```diff
@@ -39,15 +41,22 @@ const StyledSearch = styled(Input.Search)`
 `;
 
 let map: google.maps.Map;
+const mapId = '76c08a2450c223d9';
```
Comment (willmt80) [inline src/components/mapPageComponents/mapView/index.tsx:44]: ⛏️ Can we move these constants to a different file?

File: src/components/mapPageComponents/mapView/index.tsx [hunk 4]
```diff
@@ -190,29 +192,47 @@ const MapView: React.FC<MapViewProps> = ({
...
+        // Listener for when the user enters a new location
+        autocomplete.addListener('place_changed', () => {
+          const place: google.maps.places.PlaceResult = autocomplete.getPlace();
+          // If the place does not have a geometry (if the user did not enter a valid location)
           if (!place.geometry) {
-            window.alert(
-              "No details available for input: '" + place.name + "'",
-            );
+            // Predicts the place the user wanted and sets the search marker at that place
+            predictPlace(place, autoService, placesService, goToPredictedPlace);
```
Comment (willmt80) [inline src/components/mapPageComponents/mapView/index.tsx:230]: This is from Google so I assume these predictions are good, did you get a chance to test some incorrect but close addresses/names?

File: src/components/mapPageComponents/mapView/index.tsx [hunk 5]
```diff
@@ -395,9 +418,9 @@ const MapView: React.FC<MapViewProps> = ({
             let imageSize = 0;
 
             const zoomLevel = map.getZoom();
-            if (zoomLevel > 17 && zoomLevel < 20) {
+            if (zoomLevel > view + 1 && zoomLevel < STREET_ZOOM) {
```
Comment (willmt80) [inline src/components/mapPageComponents/mapView/index.tsx:421]: Is zoom level not automatically set to zoomLevel?

**Review by willmt80 — COMMENTED** [review]
Looks good! Just left a couple of questions and 1 note

File: src/components/mapPageComponents/logic/predict.ts (hunk 2 above)
Comment (SofieDunt) [reply src/components/mapPageComponents/logic/predict.ts:3]: There's documentation for `getPlacePredictions` and `getDetails` and I figured out how to use them together for automatically entering the first prediction in input (not sure if that's the question).

File: src/components/mapPageComponents/mapView/index.tsx (hunk 3 above)
Comment (SofieDunt) [reply src/components/mapPageComponents/mapView/index.tsx:44]: Would a new file under utils make sense? Or within the mapPageComponents directory?

File: src/components/mapPageComponents/mapView/index.tsx (hunk 4 above)
Comment (SofieDunt) [reply src/components/mapPageComponents/mapView/index.tsx:230]: Yeah, it works pretty well with finishing names and ignoring typos.

File: src/components/mapPageComponents/mapView/index.tsx (hunk 5 above)
Comment (SofieDunt) [reply src/components/mapPageComponents/mapView/index.tsx:421]: The default zoom level is view after clicking on a neighborhood/street zoom after searching for a place, then a user might zoom in or out. There's no broader local constant for zoom level.

File: src/components/mapPageComponents/mapView/index.tsx (hunk 3 above)
Comment (willmt80) [reply src/components/mapPageComponents/mapView/index.tsx:44]: I would say within mapPageComponents, we can make a tech debt ticket to move any numbers or other random constants for maps into it

File: src/components/mapPageComponents/constants.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { Loader } from '@googlemaps/js-api-loader';
```
Comment (willmt80) [inline src/components/mapPageComponents/constants.ts:1]: 💯

**Review by willmt80 — APPROVED** [review]
💯

**Pull request merged**